### ISSUE TITLE
Understanding sequence data in context of the pipeline

### ISSUE THREAD

**Issue by MortenHofft** [gbif/pipelines#30]
I'm trying to understand sequence data and how - if at all - it would fit into GBIF and the new pipeline. It might be a small thing that is largely ready for implementation and it might be a huge task. This is to help pipeline developers (and me) understand the scope of the problem.

Below is based on my limited understanding based on hearing @thomasstjerne and @dschigel say clever things. In conversation with Thomas we agreed that it might be nice to have a document with more details than: "it is awesome and we should do something with this".

**What could be a usage scenario**
* A user comes to GBIF with a DNA sequence. Search for it and gets the corresponding scientific name and data about where it has been seen before.
* A publisher has a dataset based on soil samples. Using a consistent reference database we match all occurrences based on their sequence to a scientific name or an OTU. This is similar to our normal latin name string matching, but using the sequence instead.

**Why would this be intersting**
* For end users
  * For end users it allows them to add a filter to only see occurrences with sequence data
  * It allows end users to search with a sequence.
* Data coverage
  * taxonomic gaps: We have gaps in fx Fungi. There is a lot of sequences based datasets for fungi that isn't shared.
  * Sampling datasets: Many of these are soil sampling based.
* For quality
  * Names can be continuously updated without the publisher needing to do anything. OTUs might get a proper name and other names might be changed. This can all be automated if there is a sequence. 
  * All names are matched against the same reference database(s) and same thresholds etc.
  * We can flag new issues (e.g. `SEQUENCE_NAME_MISMATCH`)
* For publishers
  * It shows that we care about their type of data by representing it better and considering it a quality instead of something to ignore.
  * It is slightly easier than having to do the matches themselves
  * It keeps their names updated (they might have publishers something with only an OTU, but that 2 years later is attached to a name)

**The naysayer**
Are these scenarios at all realistic? Would anyone ever use GBIF and not a specialised site to match their sequence to a species? And for publishing, wouldn't the publisher do their matching against their preferred reference data anyhow and have the scientific names that way?

**What kind of data**
It would be using marker based sequences (not shot gun) as these are small (in data amount) and widely used. Fungi seem a good testing ground as there is much sampling based data and we have gaps in fungi and would like more sample based data.

**How much data is there**
Unclear to me, but a lot I hear. What is a lot? How much data are we talking? Does it matter or is it important that we do it anyhow?

**What is the taxonomic coverage**
Fungi for sure. Insects as well. Apparently it doesn't work for plants? Animals are covered as well as I understand it?

That it would cover fungi and insects seems worth a lot in it self as these are underrepresented in GBIF.

**Who is the partners that we could work with**
Unite for fungi. Whom else?

**How to match a sequence**
How long does it take to match a string against a reference database like Unite? Does Unite have a service that we can use? Should we have our own? Can we share code? How complex is it? For Unite it seems to take around 14 seconds

**Is it at all possible to automate the matching in a confident way?**
As I understand things, given a sequence a la: 
`GAAAGTGTTATCAAGCATATGTCTTTGTATTTTAATGTGGATTTTGACAAAAACATTT...`
and knowing what markers has been used (e.g. ITS), then there is reference databases that i can ask: What is the best match for this sequence? And I will then get one or more matches and a similarity score? Something a la: `99.2% match with Acacidiplosis ananas". But is it at all possible to do an automatic matching or does it require a critical human expert to judge if this match is reasonable? This can be tried at e.g. https://unite.ut.ee/analysis.php

**Pipeline step**
During ingestion: If an occurrence has a sequence of a type we recognise (say ITS for fungi) then resolve it to a name (scientific or OTU) using our own service or from an external provider (e.g. Unite) and add that as the accepted name to the occurrence.

**How does it fit into DarwinCore?**
Is there already a natural fit in DarwinCore? The Material Sample core seems to cover the purpose. The **consensusSequence**, **barcodeSequence** and **marker** of the [amplification extension](http://rs.gbif.org/extension/ggbn/amplification.xml) are the fields of relevance.
However, GBIF currently don´t index Material Sample core type.

Does it make sense to think of above as a first step in us doing something with sequence data?

**Comment by thomasstjerne** [comment]
**Performance - a rough proof of concept**
On my laptop it took app. 1 second to build a reference database on the [UNITE dataset]:(https://plutof.ut.ee/#/datacite/10.15156%2FBIO%2F587475)
`Adding sequences from FASTA; added 30695 sequences in 1.15631 seconds.`

A single blast operation took 1 millisecond:
`yn13205:~ thomas$ date +%s`
`1517926699`
`yn13205:~ thomas$ blastn -query ~/blast/seq/test.fasta -db unite -out test.html -html
FASTA-Reader: Ignoring invalid residues at position(s): On line 1: 1-5, 7, 9, 13, 16, 21, 23, 26, 28, 32-38`
`yn13205:~ thomas$ date +%s`
`1517926700`

**Conclusion:**
It seems trivial to create a blastdb and query it, using default settings.
Performance seems good and suitable for a webservice similar that could called as a step in the ingestion.

**Comment by dschigel** [comment]
At this moment I believe that we should not have molecular identification pipeline in GBIF, but index the incoming records if they have stable OTUs (discoverable in the molecular sources of the backbone, with DOI etc). The same logic that we don't to image or sounds recognition to suggest names at GBIF. But it would very nice to have a source outside GBIF that will process georeferences and timestamped sequences before they are published through GBIF. I don't know where this will be happening, other than BOLD and UNITE. The big question how the GenBank content will be processed, not just once, but at regular intervals as both sequence content and reference databases grow dynamically. Please discuss with Donald, whose position, as I understood is that we should not index sequences, even thought there is a DwC field, but store a link out to the sequence "home", where more of molecular details can be found, also look at this example https://twitter.com/dagendresen/status/926800534857756673. I am happy to go through the rest of the list after 20 Feb.

**Comment by MortenHofft** [comment]
Thanks @dschigel - that is useful

I'm not sure what is meant with "not index sequences". So just to make sure we agree: we already have sequence data in our databases. We do however don't do anything with it - that is no processing/interpretation of it.

I guess your point is that of the 'naysayer' outlined above. And a fair point. Publishers can do the matching against their preferred reference dataset themselves. And they have the expertise to judge if it is a reasonable match.

I guess arguments for us doing it would be:
* Quality control (does name and sequence match)
* Reindexing will automatically update the name as taxonomy change. If we simply use publishers matched name, then that won't be the case.
* Consistency in the matching procedure (same thresholds and reference data)
* Make it easier for publishers (one less step for them)

**Comment by MortenHofft** [comment]
There is two parts in above sketch. 1) The name matching using sequences at ingestion time and 2) allowing users to filter for sequence data and search with sequences of their own. 2 could look like this to the end user. NB below is silly example with nonsense data

![step1](https://user-images.githubusercontent.com/6759267/35922925-27a6e244-0c1f-11e8-9269-f5d32373f58e.png)
![step2](https://user-images.githubusercontent.com/6759267/35922929-2a446d3c-0c1f-11e8-98ab-3d83031381e8.png)
![step3](https://user-images.githubusercontent.com/6759267/35922935-2c501e78-0c1f-11e8-9a5d-ad470f8de36f.png)
![step4](https://user-images.githubusercontent.com/6759267/35922938-2e3b009a-0c1f-11e8-974e-5bd21e8864d6.png)

**Comment by MortenHofft** [comment]
Below is a rough description of the current flow and how the sequence matching would fit in

We receive an occurrence a la https://api.gbif.org/v1/occurrence/1502684137/fragment
```
{
  "basisOfRecord": "MaterialSample",
  "decimalLatitude": "-14.65",
  "decimalLongitude": "145.467",
  "eventDate": "2008-09-06",
  "extensions": {
    "http://data.ggbn.org/schemas/ggbn/terms/Amplification": [
      {
        "consensusSequence": "CCTTTATCTAGTATTTGGTGCTTGAGCTGGAATAGTAGGCACAGCCTTAAGCCTTCTCATTCGAGCAGAACTAAGCCAACCTGGCGCACTCTTAGGAGACGACCAAATCTATAATGTTATTGTTACTGCACATGCCTTCGTAATGATTTTCTTTATAGTAATGCCAATTCTAATCGGGGGGTTTGGAAACTGATTAGTTCCTCTCATGCTTGGAGCCCCTGATATGGCATTCCCTCGTATGAACAACATAAGCTTCTGATTACTCCCTCCGTCATTCCTCCTTTTACTAGCTTCTTCCGGAGTTGAGGCCGGAGCCGGGACAGGTTGAACTGTCTACCCCCCACTGTCTGGTAATCTAGCCCATGCGGGAGCATCAGTAGATTTAACCATCTTCTCCCTGCACCTGGCAGGTATTTCATCAATCCTAGGAGCAATCAACTTTATCACTACCATCATCAACATAAAACCCCCCGCTATCTCTCAATACCAAACTCCTTTATTTGTTTGGGCTGTTCTAATTACTGCCGTTCTTCTACTCCTATCTCTCCCAGTCCTAGCTGCTGGCATTACTATGCTCCTGACCGACCGAAATCTTAATACTACCTTCTTCGATCCCGCAGGAGGAGGAGACCCAATTCTTTACCAACACCTC",
        "geneticAccessionNumber": "KP194104",
        "marker": "COI-5P"
      }
    ]
  },
  "phylum": "Chordata",
  "scientificName": "Pomacentrus nagasakiensis",
  ...
}
```

**Current name matching**
The pipeline then takes the string of letters `Pomacentrus nagasakiensis` and using a webservice match it against a reference dataset (the backbone) to match `Pomacentrus nagasakiensis Tanaka, 1917` with an acceptable precision. In the process it adds remarks in case of uncertainties in the match.

**Sequence matching**
Sequence matching would work the same way.

It recognise that there is a sequence and takes the string of letters `CCTTTATCTAGT...` and match it against a reference dataset (which is based on the marker e.g. BOLD for `COI` and UNITE for `ITS`). 
That could be an external service provider or it could be an internal service based on external data and matching algorithms (say UNITE data dump and BLAST matching as in https://github.com/gbif/pipelines/issues/30#issuecomment-363444340).

This returns a name that is matched to the backbone. Remarks about the transformation is added just like we do when we match to the backbone reference dataset.

**Comment by rdmpage** [comment]
Couple of thoughts. The example used by @MortenHofft (which comes from a DNA barcode dataset I uploaded as a test) has a GenBank accession number "KP194104" so one test is whether the accession number exists and matches the metadata the user supplied. This is another version of the "did the user supply a name that we recognise".

The more exciting question, and one I think @MortenHofft is suggesting is whether you can search on the sequence itself. @dschigel raises the obvious question of scope, but I think this is absolutely something GBIF should be thinking about if it is going to remain relevant in a world of sequences.

I've done some messing about with using Elastic search to index DNA sequences and essentially replicate BLAST functionality, and it looks quite doable. The trick is to index the sequence as n-grams. I'm tempted to revisit this, maybe even as a entry in the current Ebbe Nielsen Challenge...

Returning to the issue of scope, I think we're seeing a fragmentation of the sequencing landscape in biodiversity, with GenBank, BOLD, and now [GeoMe](https://www.geome-db.org), and these are often poorly interlinked with divergent notions of what metadata to collect, and what identifiers to use. This is going to make the biodiversity genomics landscape a bit messy.

I complete understand that GBIF needs to be careful about scope, but at the same time the danger is that by focussing on a classical Linnaean approach to taxonomy it will seem less relevant to a new generation of researchers. "BINs" and trying to fit those into the backbone don't really help, because a lot of the relevant information is lost by binning things (e.g., you loose phylogenetic structure both between and within "BINs").

**Comment by claudenozeres** [comment]
This from @rdmpage. I am not a phlyogeneticist, but a food web ecologist, though I often have to dig into genomics at times to discover what is I am working with because marine surveys can be incomplete/erroneous. Very grateful for GBIF, OBIS, and BOLD's BIN's because basic data is publicly available on web and R (shout out to ROpenSci for tips and packages). During searches of BINs on BOLD, I can get interesting results, but limited in what can be done as a follow-up.

I return to the question/comment that I posed to Rod at TDWG2107: it would be useful for a researcher to know/tag when a species/sequence is referenced/cited--for the researcher and the public--to enable discovery of interesting taxonomic entities (or OTU's).

Would be great for GBIF/BOLD to go further--flag names and sequences and interesting divergences beyond BINs--I have noticed literally a dozen new marine species from the Northwest Atlantic alone (or rediscovered, upon review of BHL--shout out!), but cannot speak of, because not published, nor to be published (not in my work mandate, in examining impact on stock structure of commercial fishes). Need to point to a web resource to tag, and let others carry this forward.

Regarding messy landscape. Yes. In taxonomy, for marine, we are fortunate to have WoRMS, & Species2000 for all Linnaean species? I don't work with Genbank, but have heard issues (cannot correct misidentified sequences--unlike BOLD?). Means for coordination/rectification would be great.

Yay, for Disqus for enabling continuance on an iPhylo blogpost!
